cluster-api/pkg/apis/cluster/v1alpha1/machine_types.go

```diff
@@ -107,7 +107,24 @@ type ProviderConfig struct {
 // ProviderConfigSource represents a source for the provider-specific
 // node configuration.
 type ProviderConfigSource struct {
```

Comment:
Isn't it better to inline or reference the `MachineClass` and not `ProviderConfig`?
Inline:

```yaml
apiVersion: cluster.k8s.io/v1alpha1
kind: Machine
metadata:
  name: machine-example
  namespace: baz
spec:
  class:
    value:
      capacity:
        cpu: 2
        memory: 7.5Gi
        storage: 20Gi
      allocatable:
        cpu: 2
        memory: 7.5Gi
        storage: 20Gi
      providerConfig:
        apiVersion: gceproviderconfig/v1alpha1
        kind: GCEProviderConfig
        project: GCLOUD_PROJECT_FOO
        zone: us-central1-f
        machineType: n1-standard-2
        image: projects/ubuntu-os-cloud/global/images/family/ubuntu-1604-lts
  versions:
    kubelet: 1.8.3
    containerRuntime:
      name: docker
      version: 1.12.0
  roles:
  - Node
```

Reference:

```yaml
apiVersion: cluster.k8s.io/v1alpha1
kind: MachineClass
metadata:
  name: small
  namespace: default
capacity:
  cpu: 2
  memory: 7.5Gi
  storage: 20Gi
allocatable:
  cpu: 2
  memory: 7.5Gi
  storage: 20Gi
providerConfig:
  apiVersion: gceproviderconfig/v1alpha1
  kind: GCEProviderConfig
  project: GCLOUD_PROJECT_FOO
  zone: us-central1-f
  machineType: n1-standard-2
  image: projects/ubuntu-os-cloud/global/images/family/ubuntu-1604-lts
---
apiVersion: cluster.k8s.io/v1alpha1
kind: Machine
metadata:
  name: machine-example
  namespace: baz
spec:
  class:
    ref:
      name: small
      namespace: baz
  versions:
    kubelet: 1.8.3
    containerRuntime:
      name: docker
      version: 1.12.0
  roles:
  - Node
```

I find that more flexible.